photon-api/src/main/scala/com/linkedin/photon/ml/data/InputColumnsNames.scala

```diff
@@ -20,21 +20,14 @@ package com.linkedin.photon.ml.data
  * The default values are defined in the companion object. Any of these defaults can be overridden by the user (see
  * GameParams).
  *
- * @note Tried to mix-in trait Map[A, +B], but that turned into a quagmire that is not worth it.
+ * @param array Array of input column names
  */
-class InputColumnsNames extends Serializable {
+class InputColumnsNames private (array: Array[String]) extends Serializable {
 
-  // We store default or user-defined input columns names in this efficient Array
-  private val array: Array[String] = InputColumnsNames.all.map(colName => colName.toString)
+  private val columnNames: Set[String] = array.toSet
 
-  /**
-   * Same as Map.updated, but the types are fixed.
-   *
-   * @param key The key to update
-   * @param value The value corresponding to key
-   * @return A new InputColumnsNames instance containing the update
-   */
-  def updated(key: InputColumnsNames.Value, value: String): InputColumnsNames = { array(key.id) = value; this }
+  // Check that there are no duplicate column names
+  require(array.length == columnNames.size, "Each column must have unique a name.")
```

Comment:
Why not accept `Set` instead of `Array`?

Reply:
You mean a `Map`, correct? Right now, `InputColumnNames` uses an `Array` and uses the index of the `Enum` as the key to access the column name value. We convert to a `Set` to check that the keySet is unique, and also as a convenience for the safety check on which you have commented above.

Reply:
Yeah, looks like I missed that we were using the array indices here.